File: src/legion/state/decision.py
"""Decision logic for Legion state collection.

Pure functions that determine actions based on state.
No I/O operations - just business logic.
"""

from __future__ import annotations

from legion.state import types
from legion.state.types import (
    ActionType,
    FetchedIssueData,
    IssueStatusLiteral,
    WorkerModeLiteral,
)


def suggest_action(
    status: IssueStatusLiteral | str,
    has_worker_done: bool,
    has_live_worker: bool,
    pr_is_draft: bool | None,
    has_pr: bool,
) -> ActionType:
    """Suggest action based on issue state.

    Args:
        status: Normalized issue status (IssueStatusLiteral or unknown raw value)
        has_worker_done: Whether issue has worker-done label
        has_live_worker: Whether a tmux worker session is running
        pr_is_draft: PR draft status (None if couldn't check, True if draft, False if ready)
        has_pr: Whether a PR is linked to this issue

    Returns:
        Action to take for this issue
    """
    match status:
        case types.IssueStatus.DONE:
            return "skip"

        case types.IssueStatus.TRIAGE | types.IssueStatus.ICEBOX:
            # Controller handles triage routing and Icebox pulls directly
            return "skip"

        case types.IssueStatus.BACKLOG:
            if has_worker_done:
                return "transition_to_todo"
            if has_live_worker:
                return "skip"
            return "dispatch_architect"

        case types.IssueStatus.TODO:
            if has_worker_done:
                return "transition_to_in_progress"
            if has_live_worker:
                return "skip"
            return "dispatch_planner"

        case types.IssueStatus.IN_PROGRESS:
            if has_worker_done:
                return "transition_to_needs_review"
            if has_live_worker:
                return "skip"
            return "dispatch_implementer"

        case types.IssueStatus.NEEDS_REVIEW:
            if has_worker_done:
                if not has_pr:
                    return "investigate_no_pr"  # No PR linked - needs attention
                if pr_is_draft is None:
                    return "skip"  # Has PR but couldn't check - transient, retry
                if pr_is_draft:
                    return "resume_implementer_for_changes"
                return "transition_to_retro"
            if has_live_worker:
                return "skip"
            return "dispatch_reviewer"

        case types.IssueStatus.RETRO:
            if has_worker_done:
                return "dispatch_merger"
            if has_live_worker:
                return "skip"
            return "resume_implementer_for_retro"

        case _:
            return "skip"


# Direct mapping from action to worker mode.
# Used to compute session_id for each action. "skip" uses IMPLEMENT as a
# fallback since session_id is computed even when no worker will be dispatched.
ACTION_TO_MODE: dict[ActionType, WorkerModeLiteral] = {
    "skip": types.WorkerMode.IMPLEMENT,  # fallback - not actually dispatched
    "investigate_no_pr": types.WorkerMode.IMPLEMENT,  # stuck state - needs investigation
    "dispatch_architect": types.WorkerMode.ARCHITECT,
    "dispatch_planner": types.WorkerMode.PLAN,
    "dispatch_implementer": types.WorkerMode.IMPLEMENT,
    "dispatch_reviewer": types.WorkerMode.REVIEW,
    "dispatch_merger": types.WorkerMode.MERGE,
    "resume_implementer_for_changes": types.WorkerMode.IMPLEMENT,
    "resume_implementer_for_retro": types.WorkerMode.IMPLEMENT,
    "transition_to_in_progress": types.WorkerMode.IMPLEMENT,
    "transition_to_needs_review": types.WorkerMode.REVIEW,
    "transition_to_todo": types.WorkerMode.PLAN,
    "transition_to_retro": types.WorkerMode.IMPLEMENT,
    "relay_user_feedback": types.WorkerMode.IMPLEMENT,
    "remove_worker_active_and_redispatch": types.WorkerMode.IMPLEMENT,
}


def build_issue_state(data: FetchedIssueData, team_id: str) -> types.IssueState:
    """Build final issue state with suggested action.

    Args:
        data: Fetched issue data
        team_id: Linear project UUID for session ID computation

    Returns:
        Issue state with suggested action
    """
    # Determine action based on state
    # Workers self-escalate: they post to Linear, add user-input-needed label, then exit.
    # When user responds, user-feedback-given label is added.
    if data.has_user_input_needed and data.has_user_feedback:
        # Worker asked for input and user responded - resume worker to check Linear comments
        action: ActionType = "relay_user_feedback"
    elif data.has_user_input_needed:
        # Worker asked for input but user hasn't responded yet - skip until they respond
        action = "skip"
    elif "worker-active" in data.labels and not data.has_live_worker:
        # Detect orphaned workers: worker-active label but no live window
        # Worker died - remove label and re-dispatch
        action = "remove_worker_active_and_redispatch"
    else:
        action = suggest_action(
            status=data.status,
            has_worker_done="worker-done" in data.labels,
            has_live_worker=data.has_live_worker,
            pr_is_draft=data.pr_is_draft,
            has_pr=data.has_pr,
        )

    # Compute session_id based on the action's mode
    # Default to IMPLEMENT if action not found (defensive against missing mapping)
    mode = ACTION_TO_MODE.get(action, types.WorkerMode.IMPLEMENT)
    session_id = types.compute_session_id(team_id, data.issue_id, mode)

    return types.IssueState(
        status=data.status,
        labels=data.labels,
        has_pr=data.has_pr,
        pr_is_draft=data.pr_is_draft,
        has_live_worker=data.has_live_worker,
        suggested_action=action,
        session_id=session_id,
        has_user_feedback=data.has_user_feedback,
    )


def build_collected_state(
    issues_data: list[FetchedIssueData], team_id: str
) -> types.CollectedState:
    """Build complete collected state from fetched data.

    Args:
        issues_data: List of fetched issue data
        team_id: Linear project UUID for session ID computation

    Returns:
        Complete state with all issues and suggested actions
    """
    result = types.CollectedState()

    for data in issues_data:
        result.issues[data.issue_id] = build_issue_state(data, team_id)

    return result

File: packages/daemon/src/state/decision.ts
/**
 * Decision logic for Legion state collection.
 *
 * Pure functions that determine actions based on state.
 * No I/O operations - just business logic.
 */

import {
  type ActionType,
  CiStatus,
  type CiStatusLiteral,
  type CollectedState,
  computeSessionId,
  type FetchedIssueData,
  type IssueState,
  IssueStatus,
  type IssueStatusLiteral,
  MergeableStatus,
  type MergeableStatusLiteral,
  WorkerMode,
  type WorkerModeLiteral,
} from "./types";

export function suggestAction(
  status: IssueStatusLiteral | string,
  hasWorkerDone: boolean,
  hasLiveWorker: boolean,
  prIsDraft: boolean | null,
  hasPr: boolean,
  hasTestPassed: boolean,
  ciStatus: CiStatusLiteral | null = null,
  mergeableStatus: MergeableStatusLiteral | null = null
): ActionType {
  switch (status) {
    case IssueStatus.DONE:
      return "skip";

    case IssueStatus.TRIAGE:
    case IssueStatus.ICEBOX:
      return "skip";

    case IssueStatus.BACKLOG:
      if (hasWorkerDone) {
        return "transition_to_todo";
      }
      if (hasLiveWorker) {
        return "skip";
      }
      return "dispatch_architect";

    case IssueStatus.TODO:
      if (hasWorkerDone) {
        return "transition_to_in_progress";
      }
      if (hasLiveWorker) {
        return "skip";
      }
      return "dispatch_planner";

    case IssueStatus.IN_PROGRESS:
      if (hasWorkerDone) {
        if (!hasPr) {
          return "investigate_no_pr";
        }
        if (ciStatus === CiStatus.FAILING) {
          return "resume_implementer_for_ci_failure";
        }
        if (ciStatus === CiStatus.PENDING) {
          return "retry_ci_check";
        }
        return "transition_to_testing";
      }
      if (hasLiveWorker) {
        return "skip";
      }
      return "dispatch_implementer";

    case IssueStatus.TESTING:
      if (hasWorkerDone) {
        if (hasTestPassed) {
          if (!hasPr) {
            return "investigate_no_pr";
          }
          if (ciStatus === CiStatus.FAILING) {
            return "resume_implementer_for_ci_failure";
          }
          if (ciStatus === CiStatus.PENDING) {
            return "retry_ci_check";
          }
          return "transition_to_needs_review";
        }
        return "resume_implementer_for_test_failure";
      }
      if (hasLiveWorker) {
        return "skip";
      }
      return "dispatch_tester";

    case IssueStatus.NEEDS_REVIEW:
      if (hasWorkerDone) {
        if (!hasPr) {
          return "investigate_no_pr";
        }
        if (prIsDraft === null) {
          return "retry_pr_check";
        }
        if (prIsDraft) {
          return "resume_implementer_for_changes";
        }
        // Check merge conflicts before CI status.
        // Conflicts can cause CI failures, so resolve them first.
        if (mergeableStatus === MergeableStatus.UNKNOWN) {
          return "retry_pr_check";
        }
        if (mergeableStatus === MergeableStatus.CONFLICTING) {
          return "rebase_pr";
        }
        if (ciStatus === CiStatus.FAILING) {
          return "resume_implementer_for_ci_failure";
        }
        if (ciStatus === CiStatus.PENDING) {
          return "retry_ci_check";
        }
        return "transition_to_retro";
      }
      if (hasLiveWorker) {
        return "skip";
      }
      if (hasPr && mergeableStatus === MergeableStatus.CONFLICTING) {
        return "rebase_pr";
      }
      if (hasPr && ciStatus === CiStatus.FAILING) {
        return "resume_implementer_for_ci_failure";
      }
      if (hasPr && ciStatus === CiStatus.PENDING) {
        return "retry_ci_check";
      }
      return "dispatch_reviewer";

    case IssueStatus.RETRO:
      if (hasWorkerDone) {
        return "dispatch_merger";
      }
      if (hasLiveWorker) {
        return "skip";
      }
      return "dispatch_implementer_for_retro";

    default:
      return "skip";
  }
}

export const ACTION_TO_MODE: Record<ActionType, WorkerModeLiteral> = {
  skip: WorkerMode.IMPLEMENT,
  investigate_no_pr: WorkerMode.IMPLEMENT,
  dispatch_architect: WorkerMode.ARCHITECT,
  dispatch_planner: WorkerMode.PLAN,
  dispatch_implementer: WorkerMode.IMPLEMENT,
  dispatch_implementer_for_retro: WorkerMode.IMPLEMENT,
  dispatch_reviewer: WorkerMode.REVIEW,
  dispatch_merger: WorkerMode.MERGE,
  resume_implementer_for_changes: WorkerMode.IMPLEMENT,
  resume_implementer_for_retro: WorkerMode.IMPLEMENT,
  transition_to_in_progress: WorkerMode.IMPLEMENT,
  transition_to_needs_review: WorkerMode.REVIEW,
  transition_to_todo: WorkerMode.PLAN,
  transition_to_retro: WorkerMode.IMPLEMENT,
  transition_to_done: WorkerMode.MERGE,
  relay_user_feedback: WorkerMode.IMPLEMENT,
  remove_worker_active_and_redispatch: WorkerMode.IMPLEMENT,
  add_needs_approval: WorkerMode.PLAN,
  retry_pr_check: WorkerMode.REVIEW,
  resume_implementer_for_ci_failure: WorkerMode.IMPLEMENT,
  retry_ci_check: WorkerMode.REVIEW,
  dispatch_tester: WorkerMode.TEST,
  transition_to_testing: WorkerMode.TEST,
  resume_implementer_for_test_failure: WorkerMode.IMPLEMENT,
  rebase_pr: WorkerMode.REVIEW,
};

/**
 * Modes that require server-side validation before dispatch.
 * Non-gated modes are always allowed (backward compatible).
 */
export const GATED_MODES: ReadonlySet<WorkerModeLiteral> = new Set<WorkerModeLiteral>([
  WorkerMode.MERGE,
]);

/**
 * Actions that represent actual worker dispatch (not transitions/resumes/skips).
 */
const DISPATCH_ACTIONS: ReadonlySet<string> = new Set([
  "dispatch_architect",
  "dispatch_planner",
  "dispatch_implementer",
  "dispatch_implementer_for_retro",
  "dispatch_tester",
  "dispatch_reviewer",
  "dispatch_merger",
]);

export type DispatchValidation =
  | { valid: true }
  | { valid: false; suggestedAction: ActionType; reason: string };

/**
 * Validate whether a worker mode can be dispatched given cached issue state.
 *
 * For gated modes (initially: merge), the cached suggestedAction must be a
 * dispatch action that maps to the requested mode. This prevents the controller
 * from bypassing lifecycle phases (e.g., skipping retro before merge).
 *
 * For non-gated modes, always returns valid (backward compatible).
 * For cache misses (undefined state), always returns valid (graceful degradation).
 */
export function canDispatchMode(
  cachedState: IssueState | undefined,
  requestedMode: WorkerModeLiteral
): DispatchValidation {
  if (!GATED_MODES.has(requestedMode)) {
    return { valid: true };
  }
  if (cachedState === undefined) {
    return { valid: true };
  }

  const { suggestedAction } = cachedState;
  const suggestedMode = ACTION_TO_MODE[suggestedAction];

  if (DISPATCH_ACTIONS.has(suggestedAction) && suggestedMode === requestedMode) {
    return { valid: true };
  }

  return {
    valid: false,
    suggestedAction,
    reason:
      `Cannot dispatch "${requestedMode}" worker: ` +
      `current suggestedAction is "${suggestedAction}" ` +
      `(maps to "${suggestedMode}" mode). ` +
      `The issue must reach the correct lifecycle state before "${requestedMode}" can be dispatched.`,
  };
}

const VALID_WORKER_MODES = new Set<string>([
  WorkerMode.ARCHITECT,
  WorkerMode.PLAN,
  WorkerMode.IMPLEMENT,
  WorkerMode.TEST,
  WorkerMode.REVIEW,
  WorkerMode.MERGE,
]);

export function buildIssueState(data: FetchedIssueData, legionId: string): IssueState {
  let action: ActionType;

  if (data.hasUserInputNeeded && data.hasUserFeedback) {
    action = "relay_user_feedback";
  } else if (data.hasUserInputNeeded) {
    action = "skip";
  } else if (data.labels.includes("worker-active") && !data.hasLiveWorker) {
    action = "remove_worker_active_and_redispatch";
  } else if (
    data.hasNeedsApproval &&
    data.hasHumanApproved &&
    (data.status === IssueStatus.BACKLOG || data.status === IssueStatus.TODO)
  ) {
    action = "transition_to_todo";
  } else if (
    data.hasNeedsApproval &&
    (data.status === IssueStatus.BACKLOG || data.status === IssueStatus.TODO)
  ) {
    action = "skip";
  } else {
    action = suggestAction(
      data.status,
      data.labels.includes("worker-done"),
      data.hasLiveWorker,
      data.prIsDraft,
      data.hasPr,
      data.hasTestPassed ?? false,
      data.ciStatus,
      data.mergeableStatus
    );
    if (action === "transition_to_todo") {
      action = "add_needs_approval";
    }
  }

  if (data.isBlocked === true && action.startsWith("dispatch_")) {
    action = "skip";
  }

  // Use actual worker mode for skip actions when available
  let mode: WorkerModeLiteral;
  if (
    action === "skip" &&
    data.hasLiveWorker &&
    data.workerMode &&
    VALID_WORKER_MODES.has(data.workerMode)
  ) {
    mode = data.workerMode as WorkerModeLiteral;
  } else {
    mode = ACTION_TO_MODE[action] ?? WorkerMode.IMPLEMENT;
  }
  const sessionId = computeSessionId(legionId, data.issueId, mode);

  return {
    title: data.title,
    status: data.status,
    labels: data.labels,
    hasPr: data.hasPr,
    prIsDraft: data.prIsDraft,
    ciStatus: data.ciStatus,
    mergeableStatus: data.mergeableStatus,
    hasLiveWorker: data.hasLiveWorker,
    workerMode: data.workerMode,
    workerStatus: data.workerStatus,
    suggestedAction: action,
    sessionId,
    hasUserFeedback: data.hasUserFeedback,
    isBlocked: data.isBlocked ?? false,
    source: data.source,
  };
}

export function buildCollectedState(
  issuesData: FetchedIssueData[],
  legionId: string
): CollectedState {
  const result: CollectedState = { issues: {} };

  for (const data of issuesData) {
    result.issues[data.issueId] = buildIssueState(data, legionId);
  }

  return result;
}
